src/ClientData/ScopeCollection.cpp

```diff
@@ -0,0 +1,67 @@
+
+// Copyright (c) 2022 The Orbit Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#include "ClientData/ScopeCollection.h"
+
+namespace orbit_client_data {
+
+static const ScopeStats kDefaultScopeStats;
+
+ScopeCollection::ScopeCollection(ScopeIdProvider& scope_id_provider,
+                                 std::vector<const TimerInfo*>& timers) {
+  for (const TimerInfo* timer : timers) {
+    std::optional<ScopeId> scope_id = scope_id_provider.ProvideId(*timer);
+    if (scope_id.has_value()) {
+      UpdateScopeStats(scope_id.value(), *timer);
+    }
+  }
+
+  SortTimers();
+}
+
+void ScopeCollection::UpdateScopeStats(ScopeId scope_id, const TimerInfo& timer) {
+  ScopeStats& stats = scope_stats_[scope_id];
+  const uint64_t elapsed_nanos = timer.end() - timer.start();
+  stats.UpdateStats(elapsed_nanos);
+  scope_id_to_timer_durations_[scope_id].push_back(elapsed_nanos);
+  timers_are_sorted_ = false;
+}
+
+void ScopeCollection::SetScopeStats(ScopeId scope_id, const ScopeStats stats) {
+  scope_stats_.insert_or_assign(scope_id, stats);
+}
+
+std::vector<ScopeId> ScopeCollection::GetAllProvidedScopeIds() const {
+  std::vector<ScopeId> ids;
+  std::transform(std::begin(scope_stats_), std::end(scope_stats_), std::back_inserter(ids),
+                 [](const auto& entry) { return entry.first; });
+  return ids;
+}
+
+const ScopeStats& ScopeCollection::GetScopeStatsOrDefault(ScopeId scope_id) const {
+  auto scope_stats_it = scope_stats_.find(scope_id);
+  if (scope_stats_it == scope_stats_.end()) {
+    return kDefaultScopeStats;
+  }
+  return scope_stats_it->second;
+}
+
+const std::vector<uint64_t>* ScopeCollection::GetSortedTimerDurationsForScopeId(ScopeId scope_id) {
+  if (!timers_are_sorted_) {
```

Comment:
I see 2 issues here:
1- GetSortedTimerDurations could freeze the UI.
2- If only one Id isn't sorted, this will force sorting all of them. 

I don't think that this will be particularly a problem if in the end we are only getting the durations after inserting all of them, but I will vote for either a comment to explain why this is not an issue or better a new public method `AreDurationsSorted()` and move this responsibility to the caller. 

I would like to force this method to be called ONLY after capturing/sorting somehow.

Reply:
I think, we've discussed this right when I've introduced sorting by timer durations and agreed it's better to sort the timers upon capture load. See my [comment](https://github.com/google/orbit/pull/4229/files/65064f70c6bd8745af01779000daddc7d1a52a76#r977728134)

Reply:
By adding the method I proposed (`AreDurationsSorted()`), you can check that in the caller and force to sort it.

And here you can return an empty array if they are not sorted.

At least I would like to have logged an error in this method, as we don't expect to be called with the timers not sorted.

Reply:
Instead of `AreDurationsSorted()` can I simply add a `if (timer_durations_are_sorted_) return;`to the method? And if you'd prefer I can rename it to SortTimersIfNecessary(), but an extra boolean function is an unnecessary complication for the API user.

Returning empty/nullptr here seems a bit misleading to me, but I don't feel that strongly, and at least there is a log.